Adding support for binary prefix system

maxime-bruno opened the pull request:
According to a discussion in rattan [https://github.com/stack-rs/rattan/pull/122], I propose a feature for this crate to allow the display following binary prefix system (e.g. `MiB/s`).

This can be useful for comparison with [Criterion](https://crates.io/crates/criterion) which uses this system for throughput estimation.

maxime-bruno committed:
Adding support for binary prefix system

BobAnkh reviewed (approved):
LGTM

BobAnkh commented:
Do you want to support parsing the binary system's format?

maxime-bruno committed:
Adding deserialization

maxime-bruno commented:
I'm not sure everything is perfect, especially the use of floats to move from powers of 2 to powers of 10 and vice versa.

I still have some doc issues to correct before it can be merged

maxime-bruno commented:
I also have an issue with the meaning of this comment for `fmt_<format>`
```
Enabling the display-integer feature will display integer values only

This method is preserved for backward compatibility and custom formatting.
```
```
Disabling the display-integer feature will display decimal values

This method is preserved for custom formatting.
```

I don't understand the story behind the preserved.

maxime-bruno committed:
Correcting some doc issues

maxime-bruno committed:
Adding deref traits instead of get_ref

maxime-bruno commented:
I set the `get_ref` method as deprecated to favor the usage of `Deref` traits which are designed for that

BobAnkh commented:
> I'm not sure everything is perfect, especially the use of floats to move from powers of 2 to powers of 10 and vice versa.
>
> I still have some doc issues to correct before it can be merged

Just take your time and ping me when everything is ready for a review. One more thing I want to mention is that could you please add some documentation on the units used (mainly the difference between them and the original ones). That's to say, we currently support two kinds of unit systems, so the difference (bit, byte) should be clearly clarified somewhere, I think.

> I don't understand the story behind the preserved.

This (i.e., `"9Tbps 420Gbps"`) is the original display format, presented entirely in integer form with units, which has already been used in some codebases. Considering that there may be situations where this format is preferred and to make these codebases compatible, it has been retained. However, the default format at present (`"9.42Tbps"`) is represented as float numbers, which are more intuitive, so it is recommended to set this as the default. Feel free to ping me if you still have questions about this storyline or implementation.

> I set the get_ref method as deprecated to favor the usage of Deref traits which are designed for that

Thank you for making this improvement.

maxime-bruno committed:
Adding documentation and removing float usage

maxime-bruno commented:
I solved my issue with float usage by multiplying and dividing in the right order, and with some rounding help, it's improved.
A slight issue is that decoding rounds tie to infinity when encoding does it to even (to match float behavior). Still, I don't think it is an issue as I clearly stated that due to rounding operation, we can not guarantee that decode then encode is not seamless. And cannot be perfect, switching decimal to binary is never perfect with fractional parts.

For the doc, I added a small text at the beginning of the module.

I also added support for `_` as a separator for both decoding.
And move the bandwidth logic to the bandwidth crate instead of doing it ourselves.

BobAnkh commented on Cargo.toml:
```diff
@@ -20,8 +20,10 @@ serde = { version = "1.0.130", optional = true }
 serde_json = "1.0"
 
 [features]
+default = ["binary-system"]
```
I think we shall not enable `binary-system` by default

maxime-bruno replied:
I thought so too, it's a mistake

maxime-bruno committed:
Removing `binary-system` from default features

BobAnkh commented:
I have one more issue to discuss: I think it is common to have both `bps` format and `Byte/s` format to coexist in one codebase (e.g., Rattan). Do you have any idea about this situation?

BobAnkh commented:
I mean, do we need a method to display it in binary system format? If I understand correctly, we currently display the bandwidth in `bps` format, right?

BobAnkh commented:
Oh, I was mistaken; your implementation is fine. If you don't have any further issues, I'll go ahead and merge it.

maxime-bruno commented:
I'm not sure this is that common, usually `bps` is used in network topics, like bandwidth,
and `Byte/s` is more about data throughput, like application level, and data transfer.

My idea was that criterion is displaying it in `B/s` and it seems (to the best of my knowledge) that it is a pretty decent benchmark framework. And for the test, if the reference is in `bps` in the decimal system and the result is in `B/s` in the binary system, it is not really evident to interpret the results
